src/completer.rs

```diff
@@ -0,0 +1,344 @@
+use std::collections::{BTreeMap, BTreeSet};
+use std::{rc::Rc, str::Chars};
+
+pub trait TabHandler {
+    fn handle(&mut self, line: String) -> String;
+    fn set_completer(&mut self, completer: Box<dyn Completer>);
+    fn update_line(&mut self, updated_line: String);
+    fn reset(&mut self);
+}
+
+pub struct DefaultTabHandler {
+    initial_line: String,
+    index: usize,
+    completer: Box<dyn Completer>,
+}
+impl Default for DefaultTabHandler {
+    fn default() -> Self {
+        DefaultTabHandler {
+            initial_line: String::new(),
+            index: 0,
+            completer: Box::new(DefaultCompleter::default()),
+        }
+    }
+}
+impl TabHandler for DefaultTabHandler {
+    fn handle(&mut self, line: String) -> String {
+        if self.index == 0 {
+            self.initial_line = line;
+        }
+        if let Some(s) = self.completer.complete(&self.initial_line, &mut self.index) {
+            self.index += 1;
+            s
+        } else {
+            self.initial_line.clone()
+        }
+    }
+    fn set_completer(&mut self, completer: Box<dyn Completer>) {
+        self.completer = completer;
+    }
+    fn reset(&mut self) {
+        self.index = 0;
+        self.initial_line = String::new();
+    }
+    fn update_line(&mut self, updated_line: String) {
+        if updated_line.len() < self.initial_line.len() {
+            self.initial_line = updated_line;
+            self.index = 0;
+        }
+    }
+}
+
+pub trait Completer {
+    fn complete(&self, line: &str, index: &mut usize) -> Option<String>;
+}
+
+#[derive(Debug, Clone)]
+pub struct DefaultCompleter {
+    root: CompletionNode,
+    inclusions: Rc<BTreeSet<char>>,
+    min_word_len: usize,
+}
+
+impl Default for DefaultCompleter {
+    fn default() -> Self {
+        let inclusions = Rc::new(BTreeSet::new());
+        Self {
+            root: CompletionNode::new(inclusions.clone()),
+            inclusions,
+            min_word_len: 3,
+        }
+    }
+}
+impl Completer for DefaultCompleter {
+    /// Returns an optional vector of completions based on the provided input
+    ///
+    /// # Arguments
+    ///
+    /// * `line`    The line to complete
+    /// * `index`   The candidate index
+    ///
+    /// # Example
+    /// ```
+    /// use reedline::{DefaultCompleter,Completer};
+    ///
+    /// let mut completions = DefaultCompleter::default();
+    /// completions.insert(vec!["batman","robin","batmobile","batcave","robber"].iter().map(|s| s.to_string()).collect());
+    /// assert_eq!(
+    ///     completions.complete("bat",0),
+    ///     Some("batcave".to_string()));
+    ///
+    /// assert_eq!(
+    ///     completions.complete("to the bat",0),
+    ///     None);
+    /// ```
+    fn complete(&self, line: &str, index: &mut usize) -> Option<String> {
```

Comment:
I was reading the description to this function, and I think the description is a good place to start. If you take in the line and then return a vector of results, then another part of the code could display those options to the user and they could pick one.

Here, if I'm reading it right, we have to already know what index we want?

Apologies if you explain this elsewhere.

Oh, now that I reread it, the TabHandler is the thing showing the completions to the user and here we're maybe rotating through the options?

I still think it'd make sense to have completer return all possible options, and then have the completions handler deal with the vector. That way it works with a lot of different types of completions (like popping up a table the user uses arrow keys on, for example)

Reply:
Hey @jonathandturner thanks for the comment, for me it is clear that this complete function only takes care of completing the line, how we display the options is managed in the tabhandler handle function

Reply:
@mzanrosso - something I wasn't sure of is how the tabhandler knows what the completions are. How does it know what the list is to show the user?

Reply:
I understand the point, now I managed only the inline tab so I don't need a vector but only the candidate, we should find a proper struct to do both in future

Reply:
I think returning the vector is the right call here. I was also thinking about making reedline available as a line editor for Nushell, and it'd be great to be able to use a similar competitions style, which also returns a vector of results and then the line editor handles the rest.

Reply:
Ok let me implement this change